#37653 support `default impl` for specialization

giannicic opened the pull request:
this commit implements the second part of the `default impl` feature:

>  - a `default impl` need not include all items from the trait
>  - a `default impl` alone does not mean that a type implements the trait

The first point allows rustc to compile and run something like this:

```
trait Foo {
    fn foo_one(&self) -> &'static str;
    fn foo_two(&self) -> &'static str;
}

default impl<T> Foo for T {
    fn foo_one(&self) -> &'static str {
        "generic"
    }
}

struct MyStruct;

fn  main() {
    assert!(MyStruct.foo_one() == "generic");
}
```

but it shows a proper error if trying to call `MyStruct.foo_two()`

The second point allows a `default impl` to be considered as not implementing the `Trait` if it doesn't implement all the trait items.
The tests provided (in the compile-fail section) should cover all the possible trait resolutions.
Let me know if some tests is missed.

See [referenced ](https://github.com/rust-lang/rust/issues/37653) issue for further info

r? @nikomatsakis

bors commented:
:umbrella: The latest upstream changes (presumably #45381) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
> The second point allows a default impl to be considered as not implementing the Trait if it doesn't implement all the trait items.

Hmm. My impression was that a `default impl` should never be considered as implementing the trait on its own -- regardless of whether it contains all items or not (rather like the presence of default items in a trait does not in and of itself make the trait implemented).  Does the RFC suggest otherwise, though?

nikomatsakis commented on src/librustc/traits/select.rs:
```diff
@@ -1052,12 +1052,39 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
+            // applicable to the trait predicate or the cause of the predicate is an
+            // `ObjectCastObligation`
+            if self.tcx().impl_is_default(def_id) &&
+               !self.tcx().default_impl_implement_all_methods(def_id){
```
As I wrote in a separate comment, I don't think that default impls should be considered impls -- this means we don't even need to make **candidates** for them, I think.

giannicic replied:
The RFC doesn't mention it. It was just my deduction to let them act as regular impls if they are full implemented. I'll remove it.
Concerning the candiates, do you mean moving this check when creating candiates ([select.rs#L1110](https://github.com/giannicic/rust/blob/d6cf35161444e2bb09b115d2670f485c74895d19/src/librustc/traits/select.rs#L1110)) instead of doing it as a filter at the end of the selection process?

nikomatsakis replied:
Right, so, one option might be to [filter the impls out in this loop](https://github.com/giannicic/rust/blob/d6cf35161444e2bb09b115d2670f485c74895d19/src/librustc/traits/select.rs#L1695), but really I suspect that we just don't want them to appear in the result of `for_each_relevant_impl` at all.

In general, I've been trying to think through the idea of doing a "desugaring" step to start. In other words, I'm not happy with how our trait selection code works directly on traits and impls, and I'd prefer for it to work more like chalk -- first lowering traits/impls to some kind of predicate, and then applying those.

I don't necessarily want to block support for default on making that transition, though, but in general we probably want to do a bit of refactoring to better isolate the "things we use to do trait selection" from the "actual impls" that the user wrote.

giannicic replied:
Ok, I'm working on it

nikomatsakis commented on src/librustc/ty/instance.rs:
```diff
@@ -219,6 +224,13 @@ fn resolve_associated_item<'a, 'tcx>(
             let (def_id, substs) = traits::find_associated_item(
                 tcx, trait_item, rcvr_substs, &impl_data);
+
+            check_unimplemented_trait_item(tcx,
```
This doesn't seem like the right place to be doing this check. This function runs after type-checking has completed, so this error should be reported earlier. Under what scenario did this code execute before?

giannicic replied:
At first I thought that this error should be raised in typeck but I didn't found in there where an impl method call is evaluated against a set of possible impls.
I mean, given this code 
```
 trait Foo {
     fn foo_one(&self) -> &'static str;
     fn foo_two(&self) -> &'static str;
 }
 
 struct MyStruct;
 
 default impl<T> Foo for T {
     fn foo_one(&self) -> &'static str {
         "generic"
     }
 }
 
 default impl<T: Clone> Foo for T {
     fn foo_one(&self) -> &'static str {
         "generic Clone"
     }
 }
```

If i call `MyStruct.foo_one()` the compiler select the method contained in the first impl. 
The error should be raised in case this impl doesn't contain an imlementation of `foo_one`.
Since I didn't found the selection process during typeck I've choosen to implement it after it occurs inside the `resolve_associated_item` function

Do you think I should move all inside typeck?

nikomatsakis replied:
right, so what should happen in this case is that you get an error, because `MyStruct` does not implement `Foo`. This is sort of related to the prior discussion -- having a `default impl` just means that you have some defaults that are applied *to other impls*. 

What I would expect is that if the user added a `impl Foo for MyStruct`, then we would get an error there saying that no version of `foo_two` can be found.

giannicic replied:
Ok, Got it. So there is no need to check for an unimplemented trait item. I'll remove it and include some tests on the use cases you provided.

nikomatsakis commented on src/librustc/ty/mod.rs:
```diff
@@ -994,8 +994,12 @@ impl<'a, 'gcx, 'tcx> Predicate<'tcx> {
 pub struct TraitPredicate<'tcx> {
-    pub trait_ref: TraitRef<'tcx>
+    pub trait_ref: TraitRef<'tcx>,
+    pub default_impl_check: DefaultImplCheck
```
What is the function of this field?

giannicic replied:
This field is like a flag used during Trait Resolution.
When setted to `Yes`, the `ImplCandidate` resolved from an `Obligation` that contain the `TraitPredicate` are checked to see if they refer to a `default impl`.
An `Err(Unimplemented)` is returned in that case because a `default impl` doesn't implement a `Trait`.
For example, given this code 
```
trait Foo {
    fn foo_one(&self) -> &'static str;
    fn foo_two(&self) -> &'static str;
}

default impl<T> Foo for T {
    fn foo_one(&self) -> &'static str {
        "generic"
    }
}

struct MyStruct;
```
 - a call like `MyStruct.foo_one()` should pass the trait resolution since the `TraitPredicate(MyStruct as Foo)` is marked as `No`
 - a call like `foo(MyStruct)` with foo like `fn foo<T: Foo>(x: T) -> &'static str` should fail the trait resolution since the `TraitPredicate(MyStruct as Foo)` is marked as `Yes`

nikomatsakis replied:
OK, I don't think we should need this field once we've aligned things correctly.

nikomatsakis commented:
@giannicic I wonder -- maybe we could schedule a time to chat about this change in more detail (i.e., a video chat -- or perhaps just over IRC/gitter)? I'll have to try to map out how I think it would best be implemented, I've not thought that much about it.

giannicic commented:
Sure @nikomatsakis 
I'll be on #rust IRC channel from 9:00 AM to 05:00 PM
I'm in GMT+1 time

nikomatsakis commented:
@giannicic sorry I missed you the other day. I'm at the Rust Belt Rust conference right now so my schedule is sort of wacky, but we could probalby schedule a time to "synchronize" -- actually your time zone may be helpful, since I could get up early. Perhaps Thursday (tomorrow), say at 11:00 GMT? (Which I believe is 7:00 US Eastern Time, which is my local time zone.)

giannicic commented:
@nikomatsakis Ok, see you tomorrow.

giannicic commented:
@nikomatsakis 
I've changed the `for_each_relevat_impl` method in order to not emit default impls: It's working with the use case you have provided but I've a doubt.
Some tests in the compile-fail and run-pass sections don't complete correctly. These tests have been created by me in a previous [`default impl` PR](https://github.com/rust-lang/rust/pull/37860) and basically
are taken from the related specialization tests by moving the `default `keyword from the impl items to the impl.
These tests consider default impls as regular impls but now only regular impls are emitted so, they should be deleted or changed.
I've debugged one of them [compile-fail/specialization/defaultimpl/specialization-default-projection.rs](https://github.com/rust-lang/rust/blob/b67f4283b3d04ef8e0504503123001cf642ec325/src/test/compile-fail/specialization/defaultimpl/specialization-default-projection.rs)
basically it fails when the traitpredicate `<T as Foo>` is created from the ty_projection.
Since there is no impl satisfing the `TaitPredicate `the projection fail with a selection error instead of failing with the mismatched type error
I didn't debug the others but likely they do not complete for a reason related to the `for_each_relevat_impl` change

What do you think about? Should I delete them or try to let them work by adding some additional check inside the `for_each_relevat_impl` function?

nikomatsakis commented:
@giannicic 

> These tests have been created by me in a previous default impl PR and basically
are taken from the related specialization tests by moving the defaultkeyword from the impl items to the impl.

As discussed, those tests may not be valid, since having a `default impl` is *not* equivalent to have an impl with all `default` items. So e.g. in the test you gave, I would [expect an error here that `T: Foo` may not be implemented](https://github.com/rust-lang/rust/blob/b67f4283b3d04ef8e0504503123001cf642ec325/src/test/compile-fail/specialization/defaultimpl/specialization-default-projection.rs#L27-L32). Also [here](https://github.com/rust-lang/rust/blob/b67f4283b3d04ef8e0504503123001cf642ec325/src/test/compile-fail/specialization/defaultimpl/specialization-default-projection.rs#L38). 

I was thinking: before we implement `default impl`, there are some fundamental questions that we are still wrestling with when it comes to specialization. One simple step that might be useful is to [impose the limitations that @aturon described here](https://internals.rust-lang.org/t/shipping-specialization-a-story-of-soundness/5507). If you're interested in doing that, I could open an issue and write-up some mentoring instructions (actually, I've been meaning to do that anyway).

giannicic commented:
@nikomatsakis 
Ok, I've removed them and I'm running a full test prior to commit the change.

Concerning the new task, sure I'd like to do it! 
Let me know about the new issue, in the meanwhile I'll read the linked post
Thank you

shepmaster commented:
> I was thinking: before we implement default impl, there are some fundamental questions that we are still wrestling with

@nikomatsakis I'm reading this comment as "we shouldn't do *this* PR until we figure out these other bits" — is that accurate? If so, shall we close this PR until we figure out those bits?

nikomatsakis commented:
@shepmaster @giannicic 

Sorry for radio silence. I'm not quite sure how to proceed here, to be quite honest. Part of the problem is that I am focusing ~100% on getting NLL and a few other major work items over the finish line, and it's hard for me to focus right now on thinking through what default impls would require to get implemented. That said, @giannicic, you've done some good work here, and I hate to leave you hanging!

Here is what I propose:

- I will try to read through this PR and see what to land.
- @giannicic, perhaps over next month or so you'd be interested in working on some of the looming tasks needed to push non-lexical lifetimes and the MIR-based borrowck over the finish line? There's lots of juicy stuff =)
- After that, we could revisit the question of specialization, and notably default impls? (Which I would be eager to do.)

giannicic commented:
@nikomatsakis @shepmaster 

Sure, Niko! You can rely on me ;)
Let me know when you're ready to assign something to me about those tasks. I'll read the related RFCs to get a background about the features.

Concerning the "Restrictions around lifetime dispatch" I'll file an issue with what is my understanding of the things to do so, you can review it as you find some time.

Let me know about this PR as well.
Thanks

nikomatsakis commented on src/librustc_typeck/check/wfcheck.rs:
```diff
@@ -383,8 +383,18 @@ impl<'a, 'gcx> CheckTypeWellFormedVisitor<'a, 'gcx> {
                                                   &trait_ref,
                                                   ast_trait_ref.path.span);
+
+                    // not registering predicates associcated with a `default impl`
```
OK, so, I think what this is doing is trying to skip the logic that requires a trait's where-clauses to be satisfied for it to be implemented. Which seems .. maybe right, maybe not. But what I don't know is why it is only skipping `trait` obligations, that seems definitely odd.

For example:

```rust
trait Foo<'a, T: Eq + 'a> { }

default impl<U> Foo<'static, U> for () { }
```

This will error because `U: 'static` is not known to hold, but will not error about the fact that `U: Eq` is not known to hold. Neither is obviously right to me. I suspect we should just not skip *any* checks here. Do you know what motivated you to do so?

giannicic replied:
Yes, I skip registering `default impl` obligations as a conseguence of not emitting `default impl`s inside [for_each_relevat_impl](https://github.com/giannicic/rust/blob/d6cf35161444e2bb09b115d2670f485c74895d19/src/librustc/traits/select.rs#L1695.)
At first I've just skipped all kind of obligations but I noticed that, when running the full test suite, some tests failed. 
Instead when skipping only trait obligations the tests run correctly.

If I remove the `default` keyword in your example the error given is 
```
the trait bound `U: std::cmp::Eq` is not satisfied`
```
If I understand you correctly this should be the error that I should get if i put the `default` keyword again so, I'm finding out which is the obligation involved in this check in order
to extract a general rule to follow when registering these predicates.

Perhaps I should skip only the `TraitPredicate` corresponding to the `TraitRef` of the impl. I'm doing some checks and let you know

Thanks

nikomatsakis replied:
> At first I've just skipped all kind of obligations but I noticed that, when running the full test suite, some tests failed.

Do you mean you just *always* skipped all kinds of obligations? Or only when you skipped all obligations for `default` impls? If the latter, can you tell me which tests failed?

Still, yes, I think we should be skipping *no* obligations here, and checking the same conditions on `default` impls as non-default impls. 

>  I'm finding out which is the obligation involved in this check in order
to extract a general rule to follow when registering these predicates.

I don't understand this sentence, sorry :(

giannicic commented:
@nikomatsakis , I think I've found a different solution, Hope it works for you :)

Basically I've introduced a new `InferCtxt` field called `emit_defaul_impl_candidates`. 
This field is a flag that tells to the `SelectionContext` if it can assemble `ImplCandidate`s referred to `default impl`s.
The default value of the flag is `false ` and it could be setted to `true` only when creating a new `FnCtxt ` and only when the `body_id ` field of the `FnCtxt ` is a `default impl`.
So, when [wfchecking ](https://github.com/giannicic/rust/blob/defaultimpl2/src/librustc_typeck/check/wfcheck.rs#L365) a `default impl`, the emission is allowed because the `FnCtxt.body_id` is a `default impl`, it's not allowed in all the other cases.
To let it work I've moved out the `default impl` candidate filter from [for_each_relevat_impl](https://github.com/giannicic/rust/blob/defaultimpl2/src/librustc/traits/select.rs#L1673) to [assemble_candidates_from_impls](https://github.com/giannicic/rust/blob/defaultimpl2/src/librustc/traits/select.rs#L1677) and the `SelectionContext ` [global cache is not used if the flag is true](https://github.com/giannicic/rust/blob/defaultimpl2/src/librustc/traits/select.rs#L1254).

carols10cents commented:
ping @nikomatsakis -- looks like this is back in your court!

shepmaster commented:
ping^2 @nikomatsakis -- looks like this is back in your court!

nikomatsakis commented on src/librustc/infer/mod.rs:
```diff
@@ -181,6 +181,9 @@ pub struct InferCtxt<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
     region_obligations: RefCell<Vec<(ast::NodeId, RegionObligation<'tcx>)>>,
+
+    // true if trait selection in this context should emit `default impl` candiates
+    pub emit_defaul_impl_candidates: Cell<bool>,
```
Sorry again for the delay. This flag doesn't feel right. It seems like it's true when we are checking the default impl itself, is that right? Can you give an example of the errors you were getting that you wanted to solve via this flag? 

My guess is that the right way to solve this would be by adding something into the parameter environment for a default impl; for example, an assumption that `Self: Trait<..>`, rather like we have [the trait definition itself](https://github.com/rust-lang/rust/blob/abe85ab0b232e744b602d20a65fbe92aa71c6045/src/librustc_typeck/collect.rs#L1436-L1439), which is -- after all -- a kind of default impl.

nikomatsakis replied:
Oh, I see you left a pretty detailed comment saying why you added it. =) That sounds about like what I was expecting -- but I don't think the flag is the right way to go. We would rather approach this through the parameter environment, I think.

shepmaster commented:
Heads up from triage, @giannicic — you've got some pending feedback to address.

giannicic commented:
@shepmaster
Sure! I'm working on It. I'm figuring out how to implement It using param env

qnighy commented:
Hi, how is this going? Is there something I can help?

giannicic commented:
@qnighy 
Currently I'm stucked when inserting the Self: Trait<...> predicate for a default impl as suggested by @nikomatsakis .
Since the param env is built based on the predicates of a def_id I've defined the new Self predicate inside the [explicit_predicates_of](https://github.com/giannicic/rust/blob/master/src/librustc_typeck/collect.rs#L1359) function by inserting the following code:
```
ItemImpl(_, _, defaultness, ref generics, ..) => {
    if defaultness.is_default() {
        is_impl_trait_ref = tcx.impl_trait_ref(tcx.hir.local_def_id(item.id));
        let trait_ref_node = tcx.hir.get(tcx.hir.as_local_node_id(is_impl_trait_ref.unwrap().def_id).unwrap());
        if let NodeItem(trait_ref_item) = trait_ref_node {
            if let ItemTrait(.., ref items) = trait_ref_item.node {
                let tr = ty::TraitRef {
                    def_id: is_impl_trait_ref.unwrap().def_id,
                    substs: Substs::identity_for_item(tcx, is_impl_trait_ref.unwrap().def_id)
                };
                is_trait = Some((tr, items));
            }
        }
    }
    (generics, None)
}

```

Then I've defined a new `TypeParameterDef `for Self inside the [generics_of ](https://github.com/giannicic/rust/blob/master/src/librustc_typeck/collect.rs#L837) function:

```
ItemImpl(_, _, defaultness, ref generics, ..) => {
    if defaultness.is_default()  {
        opt_self = Some(ty::TypeParameterDef {
            index: 0,
            name: keywords::SelfType.name(),
            def_id: tcx.hir.local_def_id(item.id),
            has_default: false,
            object_lifetime_default: rl::Set1::Empty,
            pure_wrt_drop: false,
            synthetic: None,
        });

        allow_defaults = true;
    }
    (generics, None)
} 
```

However, when compiling for running a test it stops during the libcore compilation with the following error:
`Region parameter out of range when substituting in region 'a (root type=Some(&'a T)) (index=0)`

It seems that the error occurs inside the  [with_fresh_ty_vars ](https://github.com/giannicic/rust/blob/master/src/librustc/traits/coherence.rs#L47) function but I still can't understand why. 

Let me know if you think I'm doing right or if you have any suggestion.
Thanks.

qnighy commented:
@giannicic thank you for the reply! I'm just an external contributor to Rust who want `default impl` to be implemented, but I hope my analysis helps.

The trigger of the error seems to be the following impl in `core::iter`.

```rust
default unsafe impl<'a, I, T: 'a> TrustedRandomAccess for Cloned<I>
    where I: TrustedRandomAccess<Item=&'a T>, T: Clone { .. }
```

which seems to intend `impl` with all items marked `default`, not really `default impl`.

And the cause seems to be `generics_of`, which is called from `rustc::ty::subst::Substs::for_item`. Your `generics_of` seems to do strange thing, because

- `opt_self` means `trait Foo<T>` has actually two type parameters because of `Self`, namely `trait Foo<Self, T>`. However, this is not the case for `impl<T> ...`. It's not actually `impl<Self, T>`.
- `allow_defaults` means type parameter defaults. This is neither the case for `impl<T> ... `.

qnighy commented:
My suggestion is as follows: what we want to do here is to consider

```rust
default impl<X> Trait<A, B> for C { .. }
```

as having bounds `C: Trait<A, B>`, namely

```rust
default impl<X> Trait<A, B> for C where C: Trait<A, B> { .. }
```

The only difference is that we have one more predicate. So I suggest to leave `generics_of` as-is and only change `explicit_predicates_of`. `explicit_predicates_of` roughly consists of two steps:

- First, use `match` to collect information about the node.
- Second, collect predicates according to the information.

In the second step, the current sources of predicates are:

- Supertrait bounds (`Self: SuperTrait<..>` where `trait SubTrait : SuperTrait<..>`)
- The trait itself (`Self: SubTrait` where `trait SubTrait : SuperTrait<..>`)
- Inline predicates on lifetime parameters (`'a: 'b` where `fn f<'a: 'b, 'a>() {}`)
- Inline predicates on type parameters (`T: Foo` where `struct X<T: Foo>();`)
- Predicates in the `where` clause (`T: Debug` where `enum Error<T> where T: Debug {}`)
- Predicates from associated types (`Self::Target: Sized` where `trait Foo { type Target; }`)
- Predicates from universal `impl Trait`s (`<impl Fn()@0>: Fn()` where `fn f(f: impl Fn()) {}`)

It seems natural to add a new source to the above list. To do this, we can just collect TraitRef and Self type from the `default impl` in the first step and transform them into a predicate using `AstConv::ast_ty_to_ty`, `AstConv::instantiate_mono_trait_ref`, and `ToPredicate::to_predicate` in the second step.

shepmaster commented:
Ping from triage, @giannicic! Did any of @qnighy's suggestions help at all?

giannicic commented:
@shepmaster 
I'm following the second suggestion. I'll let you know if it works.

giannicic commented:
@qnighy 
If I remove the code inside `generics_of` and add a `Self:Trait` predicate inside  `explicit_predicates_of` the error returned is: 
`Type parameter `Self/#0` (Self/0) out of range when substituting (root type=Some(Self)) substs=['_#0r, _, _]`

It's still inside the lib core function and I've already seen it, It's the reason why I've changed the `generics_of` by including a `TypeParameterDef`.
Basically before there was a region parameter error and now there is a type parameter error.
Probably the predicate is not included correctly.... I'm still figuring out how to proper do this change. :(

qnighy commented:
@giannicic Could you show me the code (e.g. by pointing to a branch)?

giannicic commented:
@qnighy 
Well, basically the changes are all inside the explicit_predicates_of function and are the following

```
let mut is_trait = None;
// giannicic: New variable
let mut is_impl_trait = None;

let icx = ItemCtxt::new(tcx, def_id);
let no_generics = hir::Generics::empty();
let (ast_generics, opt_inputs) = match node {
    NodeTraitItem(item) => {
        match item.node {
            TraitItemKind::Method(ref sig, _) => (&item.generics, Some(&sig.decl.inputs)),
            _ => (&item.generics, None)
        }
    }

    NodeImplItem(item) => {
        match item.node {
            ImplItemKind::Method(ref sig, _) => (&item.generics, Some(&sig.decl.inputs)),
            _ => (&item.generics, None)
        }
    }

    NodeItem(item) => {
        match item.node {
            ItemFn(ref decl, .., ref generics, _) => (generics, Some(&decl.inputs)),

            ItemImpl(_, _, defaultness, ref generics, ..) => {
                // giannicic: I've added this block to get the TraitRef
                if defaultness.is_default() {
                    let is_impl_trait_ref = tcx.impl_trait_ref(tcx.hir.local_def_id(item.id));
                    is_impl_trait = Some(ty::TraitRef {
                        def_id: is_impl_trait_ref.unwrap().def_id,
                        substs: Substs::identity_for_item(tcx, is_impl_trait_ref.unwrap().def_id)
                    });
                }
                (generics, None)
            }
            ItemTy(_, ref generics) |
            ItemEnum(_, ref generics) |
            ItemStruct(_, ref generics) |
            ItemUnion(_, ref generics) => {
                (generics, None)
            }

            ItemTrait(_, _, ref generics, .., ref items) => {
                is_trait = Some((ty::TraitRef {
                    def_id,
                    substs: Substs::identity_for_item(tcx, def_id)
                }, items));
                (generics, None)
            }

            _ => (&no_generics, None)
        }
    }

    NodeForeignItem(item) => {
        match item.node {
            ForeignItemStatic(..) => (&no_generics, None),
            ForeignItemFn(ref decl, _, ref generics) => (generics, Some(&decl.inputs)),
            ForeignItemType => (&no_generics, None),
        }
    }

    NodeTy(&Ty { node: TyImplTraitExistential(ref exist_ty, _), span, .. }) => {
        let substs = Substs::identity_for_item(tcx, def_id);
        let anon_ty = tcx.mk_anon(def_id, substs);

        debug!("explicit_predicates_of: anon_ty={:?}", anon_ty);

        // Collect the bounds, i.e. the `A+B+'c` in `impl A+B+'c`.
        let bounds = compute_bounds(&icx,
                                    anon_ty,
                                    &exist_ty.bounds,
                                    SizedByDefault::Yes,
                                    span);

        debug!("explicit_predicates_of: bounds={:?}", bounds);

        let predicates = bounds.predicates(tcx, anon_ty);

        debug!("explicit_predicates_of: predicates={:?}", predicates);

        return ty::GenericPredicates {
            parent: None,
            predicates: predicates
        };
    }

    _ => (&no_generics, None)
};

let generics = tcx.generics_of(def_id);
let parent_count = generics.parent_count() as u32;
let has_own_self = generics.has_self && parent_count == 0;

let mut predicates = vec![];

// Below we'll consider the bounds on the type parameters (including `Self`)
// and the explicit where-clauses, but to get the full set of predicates
// on a trait we need to add in the supertrait bounds and bounds found on
// associated types.
if let Some((trait_ref, _)) = is_trait {
    predicates = tcx.super_predicates_of(def_id).predicates;

    // Add in a predicate that `Self:Trait` (where `Trait` is the
    // current trait).  This is needed for builtin bounds.
    predicates.push(trait_ref.to_poly_trait_ref().to_predicate());
}

// giannicic: And I've used the TraitRef to insert the predicate
if let Some(trait_ref) = is_impl_trait {
    predicates.push(trait_ref.to_poly_trait_ref().to_predicate());
}
```

qnighy commented:
@giannicic Replacing
```rust
let is_impl_trait_ref = tcx.impl_trait_ref(tcx.hir.local_def_id(item.id));
is_impl_trait = Some(ty::TraitRef {
    def_id: is_impl_trait_ref.unwrap().def_id,
    substs: Substs::identity_for_item(tcx, is_impl_trait_ref.unwrap().def_id)
});
```
with
```rust
is_impl_trait = tcx.impl_trait_ref(tcx.hir.local_def_id(item.id));
```
seems to work.

Altering `substs` by `Substs::identity_for_item` seems to do the following: for example, suppose we have a default impl
```rust
default impl<X, Y> Foo<Box<Y>> for Vec<X> { .. }
```
Then `tcx.impl_trait_ref` gives `Vec<X>: Foo<Box<Y>>` predicate. If we alter its `substs` by identity, it turns into `X: Foo<Y>`. What we want here is the former one.

giannicic commented:
@qnighy 
Have you successfully compiled libcore with that change?
Here on my PC the previuos error has gone but I get another one, probably related to the newly inserted  predicate:

```
error[E0308]: mismatched types
   --> src/libcore/iter/mod.rs:585:9
    |
584 |     unsafe fn get_unchecked(&mut self, i: usize) -> Self::Item {
    |                                                     ---------- expected `<iter::Cloned<I> as iter::iterator::Iterator>::Item` because of return type
585 |         self.it.get_unchecked(i).clone()
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected associated type, found type parameter
    |
    = note: expected type `<iter::Cloned<I> as iter::iterator::Iterator>::Item`
               found type `T`
```

qnighy commented:
@giannicic Hm, there seem to be two causes of the problem:

1. `default unsafe impl` should have been `unsafe impl` with the two items marked `default`, as it intends to actually implement `TrustedRandomAccess`. I already fixed it in my codebase, so I didn't notice the one below.
2. However, even if it is `default impl`, it's strange that it doesn't compile. I suspect that we have two sources of impl for `<Cloned<I> as Iterator>::Item` during typechecking -- from the trait bound `Cloned<I>: TrustedRandomAccess` we just added and from the existing impl `Iterator for Cloned<I>`. According to winnowing priorities in trait selection, the former will be selected. However, we want the latter (or other concrete one) to know its associated type.

So my suggestion is to fix the former issue and postpone the latter issue, but we might better ask reviewers. @nikomatsakis what do you think about it?

nikomatsakis commented:
Sorry, I've not had a chance to weigh in yet. I keep running out of time before I get to reviewing this PR. I'm scheduling a lot to take a look for Friday, hopefully I can get to it earlier.

nikomatsakis commented on src/librustc_typeck/check/mod.rs:
```diff
@@ -1729,6 +1729,8 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                param_env: ty::ParamEnv<'tcx>,
                body_id: ast::NodeId)
                -> FnCtxt<'a, 'gcx, 'tcx> {
+        FnCtxt::set_emit_default_impl_candidates(inh, body_id);
```
What is the goal of *this*? I guess it's meant to allow the `default impl` to "assume itself", as it were? But I think we should handle that instead by having the environment for a `default impl` include itself. In other words:

```rust
default impl<T: Debug> Foo for T {
    fn bar() {
        // Environment would include not only `T: Debug` but also `T: Foo`
    }
}
```

This version promises a bit too much, from what I can tell. For example, I think that this would type-check:

```rust
trait Foo { }
default impl<T> Foo for T { }

trait Bar { fn bar(); }
default impl<T> Bar for T { 
    fn bar() {
        T::foo(); // only provided through a `default impl`
    }
}
```

nikomatsakis commented:
I'm sorry again for the long delay! Trying to stay on top of things better, just had a very long backlog of things to catch up on after the impl period.

qnighy commented:
@giannicic I suppose that this PR is not up-to-date with your codebase because of the error you're getting. I suggest you to push your latest code, with `default unsafe impl TrustedRandomAccess` changed to `unsafe impl` with `default fn`s. That would tentatively fix the error, probably.

giannicic commented:
Yes, exactly. The code is not updated to my latest. My attempt to do what @nikomatsakis has suggested (to include appropriate predicates in the param_env) is failing right now due to the things discussed in the previous posts with @qnighy so I didn't commit them.
Currently I'm following a different approach which is inserting the predicate only inside the [param_env](https://github.com/giannicic/rust/blob/master/src/librustc/ty/mod.rs#L2577) function. Maybe limiting the `default impl` predicate insertion only when the param_env is built is the right way to procede

nikomatsakis commented:
@giannicic oh, sorry, that makes sense. In general, though, please push the latest context, even if it fails! That will help me to understand what's going wrong. =)

nikomatsakis commented:
@giannicic you planning on pushing?

giannicic commented:
@nikomatsakis not yet because It seems that inserting the `Self: Trait<...>` predicate only inside the `param_env` function has worked. Now I'm finishing the PR by properly filter out any `default impl`s inside the `assemble_candidates_from_impl` function. Then I think i should do a rebase so problaby in a week i'll push it all.

nikomatsakis commented:
@giannicic ah ok =)

giannicic commented:
@nikomatsakis 
As per your [request](https://github.com/rust-lang/rust/pull/45404#discussion_r155090310) I've removed the flag and added a `Self: Trait<..>` inside the `param_env` when wfchecking a `default impl`. This predicate is used in place of the flag to decide if the [assemble_candidates_from_impls](https://github.com/giannicic/rust/blob/defaultimpl2/src/librustc/traits/select.rs#L1706) should emit default impls.
Let me know if something is wrong
Thanks!

nikomatsakis commented:
@giannicic cool! will look asap

nikomatsakis commented on src/librustc/traits/select.rs:
```diff
@@ -1710,22 +1710,44 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
+        // Check if default impls should be emitted.
+        // default impls are emitted if the param_env is refered to a default impl.
+        // The param_env should contain a Self: Trait<..> predicate in those cases
+        let self_trait_is_present:Vec<&ty::Predicate<'tcx>> =
```
Hmm, I don't think we should have to modify select.rs at all. Instead, what I would expect is for us to modify the set of predicates associated with default impls, analogously to what we do with traits. This would be part of the `predicates_of` query -- for example, in traits, we insert `Self: Trait<...>` right here:

https://github.com/rust-lang/rust/blob/4d2d3fc5dadf894a8ad709a5860a549f2c0b1032/src/librustc_typeck/collect.rs#L1444-L1446

I would think that we would want to do the same code in the case of a `default impl`. In that case, I don't think we should have to modify `select.rs` at all. The required assumptions will be found in the environment.

giannicic commented:
@nikomatsakis 
Ok, now the trait is inserted inside the `predicate_of` fn. 
As suggested by @qnighy i've changed `default unsafe impl TrustedRandomAccess` by removing the `default` keyword from the `impl` and inserting it in the `fn`s.
It seems to work now,
Thank you all for your patience :)

nikomatsakis commented on src/libcore/iter/mod.rs:
```diff
@@ -579,15 +579,15 @@ impl<'a, I, T: 'a> FusedIterator for Cloned<I>
 
 #[doc(hidden)]
-default unsafe impl<'a, I, T: 'a> TrustedRandomAccess for Cloned<I>
+unsafe impl<'a, I, T: 'a> TrustedRandomAccess for Cloned<I>
```
Indeed, I think `default` here was incorrect. 

cc @rust-lang/libs -- `TrustedRandomAccess` was *meant* to be implemented for `Cloned<I>`, right?

(i.e., a default impl is not itself an impl, it just supplies defaults to other impls)

alexcrichton replied:
Yeah I think this looks fine by me!

nikomatsakis commented on src/librustc_typeck/collect.rs:
```diff
@@ -1446,6 +1452,10 @@ fn explicit_predicates_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
         predicates.push(trait_ref.to_poly_trait_ref().to_predicate());
     }
 
+    if let Some(trait_ref) = is_default_impl_trait {
```
Can you add a comment here? Something like this:

---

```
// In default impls, we can assume that the self type implements
// the trait. So in:
//
//     default impl Foo for Bar { .. }
//
// we add a default where clause `Foo: Bar`. We do a similar thing for traits
// (see below). Recall that a default impl is not itself an impl, but rather a
// set of defaults that can be incorporated into another impl.
```

nikomatsakis commented on src/test/compile-fail/specialization/defaultimpl/specialization-trait-item-not-implemented.rs:
```diff
@@ -0,0 +1,31 @@
+}
+
+impl Foo for MyStruct {}
+//~^ ERROR not all trait items implemented, missing: `foo_two` [E0046]
```
👍

nikomatsakis commented on src/test/compile-fail/specialization/defaultimpl/specialization-trait-not-implemented.rs:
```diff
@@ -0,0 +1,30 @@
+struct MyStruct;
+
+default impl<T> Foo for T {
+    fn foo_one(&self) -> &'static str {
```
👍

nikomatsakis commented on src/test/compile-fail/specialization/defaultimpl/specialization-trait-not-implemented.rs:
```diff
@@ -0,0 +1,30 @@
+// except according to those terms.
+
+#![feature(specialization)]
+
```
It'd be nice to have comments on these tests with what they are testing:

---

Tests that (a) default impls do not have to supply all items and (b) a default impl does not count as an impl (in this case, an incomplete default impl).

nikomatsakis commented on src/test/compile-fail/specialization/defaultimpl/specialization-trait-item-not-implemented.rs:
```diff
@@ -0,0 +1,31 @@
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
```
It'd be nice to have comments on these tests with what they are testing:

---

Tests that (a) default impls do not have to supply all items but (b) regular impls do.

nikomatsakis commented on src/test/compile-fail/specialization/defaultimpl/specialization-wfcheck.rs:
```diff
@@ -10,23 +10,9 @@
 
-// Regression test for ICE when combining specialized associated types and type
-// aliases
+trait Foo<'a, T: Eq + 'a> { }
```
Comment:

---

Tests that a default impl still has to have a WF trait ref.

nikomatsakis commented on src/test/run-pass/specialization/defaultimpl/specialization-trait-item-not-implemented.rs:
```diff
@@ -8,30 +8,27 @@
-// Test that non-method associated functions can be specialized
-
 #![feature(specialization)]
 
```
Comment:

---

Tests that we can combine a default impl that supplies one method with a full impl that supplies the other, and they can invoke one another.

nikomatsakis commented on src/test/run-pass/specialization/defaultimpl/specialization-trait-item-not-implemented.rs:
```diff
@@ -8,30 +8,27 @@
+
+default impl<T> Foo for T {
+    fn foo_one(&self) -> &'static str {
+        "generic"
```
Maybe nice to have this invoke a `foo_three` that the regular impl supplies, just to complete the "calling back and forth" cycle.

nikomatsakis commented on src/test/run-pass/specialization/defaultimpl/default-methods.rs:
```diff
@@ -1,94 +0,0 @@
-//     $i32   &str  $Vec<T>
-//                    /\
-//                   /  \
-//            Vec<i32>  $Vec<i64>
```
What's up with these tests? Were they just not quite correct? I like that they are testing more advanced scenarios, but it seems fine to remove them from now and revisit them later. I've not looked closely at them mind you. =)

giannicic replied:
This test has been created by me in a previous PR by just coping the files that was in the specialization dir. As discussed here https://github.com/rust-lang/rust/pull/45404#issuecomment-340310164, i think that it's wrong because it's on the assumption that a default impl is also a regular impl so you've removed it.

giannicic commented:
@nikomatsakis 
I've inserted the comments you suggested and changed the run-pass test.
Thanks

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 220bb22 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 220bb22e1b621ad5a10a44080e3e1872d99f3e9f -->

bors commented:
:hourglass: Testing commit 220bb22e1b621ad5a10a44080e3e1872d99f3e9f with merge efda9bae8788fdc68ea1d05be3c5b66d3e291961...

nikomatsakis commented:
BTW, @giannicic and @qnighy -- check out [this internals thread announcing the traits working group](https://internals.rust-lang.org/t/announcing-traits-working-group/6804), I thought you folks might be interested.
